### PR TITLE
Handled adherent changing password

### PR DESCRIPTION
**Pull request by HeahDude**
Closes [EM-23](http://jira.sensiolabs.net/browse/EM-23).

### CODE REVIEW

File: app/Resources/views/adherent/change_password.html.twig [hunk 1]
```diff
@@ -4,4 +4,22 @@
...
+
+        {{ form_label(form.old_password, 'Mot de passe actuel') }}
+        {{ form_errors(form.old_password) }}
+        {{ form_widget(form.old_password, { attr: { placeholder: 'Entrez votre mot de passe actuel' } }) }}
+
+        {{ form_label(form.password.first, 'Nouveau mot de passe') }}
+        {{ form_errors(form.password.first) }}
+        {{ form_widget(form.password.first, { attr: { placeholder: 'Entrez votre nouveau mot de passe' } }) }}
+
+        {{ form_label(form.password.second, 'Nouveau mot de passe une deuxième fois') }}
```
Comment (romain-pierre) [inline app/Resources/views/adherent/change_password.html.twig]: Je vois plus souvent "Confirmation", voir même aucun label pour ce champs.

File: src/AppBundle/Membership/AdherentChangePasswordHandler.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+use AppBundle\Entity\Adherent;
+use Doctrine\Common\Persistence\ObjectManager;
+use Symfony\Component\Security\Core\Encoder\EncoderFactoryInterface;
+
+class AdherentChangePasswordHandler
+{
+    private $manager;
+    private $encoderFactory;
+
+    public function __construct(ObjectManager $manager, EncoderFactoryInterface $encoderFactory)
```
Comment (romain-pierre) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: Why using EncoderFactory instead of PasswordEncoder?

File: tests/AppBundle/Controller/AdherentControllerTest.php [hunk 3]
```diff
@@ -207,6 +207,50 @@ public function testEditAdherentInterests()
...
+    {
+        $this->authenticateAsAdherent($this->client, 'carl999@example.fr', 'secret!12345');
+
+        $crawler = $this->client->request(Request::METHOD_GET, '/espace-adherent/mon-profil/changer-mot-de-passe');
+
+        $this->assertCount(1, $crawler->filter('input[name="adherent_change_password[old_password]"]'));
+        $this->assertCount(1, $crawler->filter('input[name="adherent_change_password[password][first]"]'));
+        $this->assertCount(1, $crawler->filter('input[name="adherent_change_password[password][second]"]'));
+
+        // Submit the profile form with invalid data
```
Comment (romain-pierre) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Please can you add for this step (invalid data) an assert to be sure that the password have not been updated? With `$this->authenticateAsAdherent')` or directly from the database?

File: app/Resources/views/adherent/change_password.html.twig (hunk 1 above)
Comment (HeahDude) [inline app/Resources/views/adherent/change_password.html.twig]: I agree, but cf user stories

File: src/AppBundle/Membership/AdherentChangePasswordHandler.php (hunk 2 above)
Comment (HeahDude) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: Because the factory already holds every instances, no need to create an extra service as done in other places.

File: src/AppBundle/Entity/Adherent.php [hunk 4]
```diff
@@ -221,6 +221,11 @@ public function getActivatedAt()
         return $this->activatedAt;
     }
 
+    public function changePassword($newPassword)
```
Comment (tgalopin) [inline src/AppBundle/Entity/Adherent.php]: String typehint?

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 3 above)
Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: How could it be possible? It's the same client redirected that we're handling all along the test here.

File: src/AppBundle/Entity/Adherent.php (hunk 4 above)
Comment (HeahDude) [inline src/AppBundle/Entity/Adherent.php]: Good catch, thx.

Comment (HeahDude) [inline src/AppBundle/Entity/Adherent.php]: Fixed.

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 3 above)
Comment (romain-pierre) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: The client uses a cookie to keep its connection state. So, even if your password is updated (in the database), you will be able to perform some new requests. Ok, if the password is updated, you will be unable to make the password change in the next step...

What do you think about splitting your test in parts:
- Submit empty form
- Submit valid new passwords but wrong old
- Submit valid old password but not same new ones
- Submit valid form

So, just add two steps in your case. This is a security feature, this is why I like to have wide tests.

File: src/AppBundle/Membership/AdherentChangePasswordHandler.php (hunk 2 above)
Comment (romain-pierre) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: Ok, its up to you. Personally I prefer to use the `PasswordEncoder` (available at `security.password_encoder`) because it's just do what you do (so more convenient) :)

Comment (HeahDude) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: The service you mention just calls the factory, so no need for this level of indirection IMHO (see https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Resources/config/security.xml#L53).

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 3 above)
Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: I don't understand the need, here the test have a response with errors (due to empty string), how could the password change then (error means no persistence possible). Testing otherwise would mean we would test symfony flow, weird.

File: app/Resources/translations/messages.fr.yml [hunk 5]
```diff
@@ -30,6 +30,7 @@ adherent.reset_password.expired_key: Votre demande de réinitialisation de mot d
 
 adherent.update_profile.success: Vos informations ont été mises à jour avec succès.
 adherent.update_interests.success: Vos centres d'intérêts ont bien été mis à jour.
+adherent.update_password.success: Vos centres d'intérêts ont bien été mis à jour.
```
Comment (hhamon) [inline app/Resources/translations/messages.fr.yml]: Error message does not message the key name ^^

File: src/AppBundle/Form/AdherentChangePasswordType.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Security\Core\Validator\Constraints\UserPassword;
+
+class AdherentChangePasswordType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->add('old_password', PasswordType::class, [
+            'mapped' => false,
+            'constraints' => new UserPassword(['message' => 'adherent.wrong_password']),
```
Comment (hhamon) [inline src/AppBundle/Form/AdherentChangePasswordType.php]: Missing `NotBlank` constraint as well?

File: src/AppBundle/Membership/AdherentChangePasswordHandler.php (hunk 2 above)
Comment (hhamon) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: Well I agree with Romain as we already have the user so we can use the `security.password_encoder` service and inject the `UserPasswordEncoderInterface` service. Probably easier to mock in unit tests.

File: app/Resources/translations/messages.fr.yml (hunk 5 above)
Comment (HeahDude) [inline app/Resources/translations/messages.fr.yml]: Good catch, thanks!

File: src/AppBundle/Form/AdherentChangePasswordType.php (hunk 6 above)
Comment (HeahDude) [inline src/AppBundle/Form/AdherentChangePasswordType.php]: Not needed with this one.

File: src/AppBundle/Membership/AdherentChangePasswordHandler.php (hunk 2 above)
Comment (HeahDude) [inline src/AppBundle/Membership/AdherentChangePasswordHandler.php]: Fixed.

File: app/Resources/translations/messages.fr.yml (hunk 5 above)
Comment (HeahDude) [inline app/Resources/translations/messages.fr.yml]: Fixed.
